### PR TITLE
Feature: Implement Budget Handling.

### PR DESCRIPTION
**Pull request by cbuchli**

### CODE REVIEW

**Comment by cbuchli** [comment]
@mweibel not sure if ready, will review it myself tomorrow again.

File: src/Classes/Controller/ApiJobExecuteController.php [hunk 1]
```diff
@@ -171,29 +171,39 @@ public function execAction(): ResponseInterface
...
 
             // run the job and check if the replicas have changed
+            // TODO: Fix this ugly mess and cleanup DispatchConfig all together
             $previousReplicaCount = JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->getDispatchConfig()->getReplicaCountForJob($this->job);
-            JobFactory::getInstanceOfJob($this->job, $this->execution)->$command(json_decode((string) $this->request->getBody(), true) ?: []);
+            JobFactory::getInstanceOfJob($this->job, $this->execution)->$command(json_decode((string)$this->request->getBody(), true) ?: []);
             $newReplicaCount = JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->getDispatchConfig()->getReplicaCountForJob($this->job);
 
+            if (!JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->isExecutionStillAffordable()) {
+                throw new RuntimeException('Budget is used up!');
```
Comment (mweibel) [inline src/Classes/Controller/ApiJobExecuteController.php]: I think we should instead render a specific status code and log it. Maybe also send a mail to the owner?

File: src/Classes/Controller/ApiJobExecuteController.php [hunk 2]
```diff
@@ -171,29 +171,39 @@ public function execAction(): ResponseInterface
...
-            return $this->render(array_merge([
+            $result = [
                 'status' => 'success',
                 'id' => $this->execution->getId(),
-            ], $estimates));
+            ];
+            if ($command === 'run') {
+                $result['estimates'] = JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->getExecutionEstimates();
+            } else {
+                $this->job->setBudgetUsed(JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->getExecutionEstimates()['costs'] + $this->job->getBudgetUsed());
```
Comment (mweibel) [inline src/Classes/Controller/ApiJobExecuteController.php]: else case is `delete` of the execution, right? 
Wouldn't this trigger 2 times a reduction of budget when also submitresult is called?

File: src/Classes/Job/AbstractExecute.php [hunk 3]
```diff
@@ -54,11 +54,11 @@ public function __construct(Job $job, Execution $execution = null)
     public function create(array $jobObject): bool
     {
         $this->job->setName($jobObject['name'] ?? 'Automatically named during add')
-            ->setCpus((int) ($jobObject['cpus'] ?? 0))
```
Comment (mweibel) [inline src/Classes/Job/AbstractExecute.php]: did you run cs fixer on this PR?

File: tests/Integration/CostCalculatorTest.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    public function testEstimatesAreBasicallyCorrect(): void
+    {
+        $job = (new Job())->setType(JobType::BUSYBOX)->setOwner($this->user)->setStatus(JobStatus::READY)->setBudget(10);
+        $this->infrastructure->getEntityManager()->persist($job);
+        $this->infrastructure->getEntityManager()->flush();
+        $tokenHeader = ['Authorization' => 'Bearer ' . JwtUtility::generateToken(null, $this->user)['token']];
+        $result = json_decode($this->runWebApp('POST', '/api/job/' . $job->getId() . '/execute', true, $tokenHeader)->getBody(), true);
+        $this->assertArrayHasKey('estimates', $result);
+
+        $this->assertEquals(1030, $result['estimates']['duration']);
```
Comment (mweibel) [inline tests/Integration/CostCalculatorTest.php]: why 1030?

File: tests/Integration/CostCalculatorTest.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+        $comletedTimestamp = $result['estimates']['completion'];
+        $now = (new \DateTime('now', ServerUtility::getTimezoneObject()))->getTimestamp();
+        $this->assertLessThan($now + 1030 + 5, $comletedTimestamp);
+        $this->assertGreaterThan($now + 1030 - 5, $comletedTimestamp);
```
Comment (mweibel) [inline tests/Integration/CostCalculatorTest.php]: why +/- 5?

File: tests/Integration/CostCalculatorTest.php (hunk 4 above)
Comment (mweibel) [reply tests/Integration/CostCalculatorTest.php]: aha, I see why, nevermind.

File: tests/Integration/CostCalculatorTest.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        $job = (new Job())->setType(JobType::BUSYBOX)->setOwner($this->user)->setStatus(JobStatus::READY)->setBudget(10);
+        $this->infrastructure->getEntityManager()->persist($job);
+        $this->infrastructure->getEntityManager()->flush();
+        $tokenHeader = ['Authorization' => 'Bearer ' . JwtUtility::generateToken(null, $this->user)['token']];
+        $result = json_decode($this->runWebApp('POST', '/api/job/' . $job->getId() . '/execute', true, $tokenHeader, ['limit'=>10])->getBody(), true);
+        $this->assertArrayHasKey('estimates', $result);
+
+        $this->assertEquals(130, $result['estimates']['duration']);
+
+        $comletedTimestamp = $result['estimates']['completion'];
```
Comment (mweibel) [inline tests/Integration/CostCalculatorTest.php]: typo: `comleted` -> `completed`

Comment (mweibel) [reply tests/Integration/CostCalculatorTest.php]: also in the other tests, actually :D

File: tests/Integration/CostCalculatorTest.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        $job = (new Job())->setType(JobType::BUSYBOX)->setOwner($this->user)->setStatus(JobStatus::READY)->setBudget(1);
+        $this->infrastructure->getEntityManager()->persist($job);
+        $this->infrastructure->getEntityManager()->flush();
+        $tokenHeader = ['Authorization' => 'Bearer ' . JwtUtility::generateToken(null, $this->user)['token']];
+
+        $result = $this->runWebApp('POST', '/api/job/' . $job->getId() . '/execute', true, $tokenHeader);
+        $this->assertEquals(200, $result->getStatusCode());
+        $result = $this->runWebApp('POST', '/api/job/' . $job->getId() . '/execute', true, $tokenHeader, ['limit' => 10000]);
+        $this->assertEquals(500, $result->getStatusCode());
```
Comment (mweibel) [inline tests/Integration/CostCalculatorTest.php]: shouldn't throw a 500 as it's not an internal server error.

File: src/Classes/Controller/ApiJobExecuteController.php (hunk 2 above)
Comment (cbuchli) [reply src/Classes/Controller/ApiJobExecuteController.php]: yes it would, but like this, people can execute stuff for free when they never call submitresult and then delete the execution ;)

However, you're right, this is too sketchy

File: src/Classes/Controller/ApiJobExecuteController.php (hunk 1 above)
Comment (cbuchli) [reply src/Classes/Controller/ApiJobExecuteController.php]: Yes, defenitely. thanks

File: tests/Integration/CostCalculatorTest.php (hunk 5 above)
Comment (cbuchli) [reply tests/Integration/CostCalculatorTest.php]: In case the test runs at exactly x:y:z.49, then the estimated completion is 1s off. so equals() doesn't always work... idk a better solution, do you? 

(of course we could also just do 1s split)

**Comment by cbuchli** [comment]
> I'm unsure if we should go forward with this, honestly. IMHO we shouldn't rush this feature.
> One thing which I think is necessary: The estimated cost of each execution should be stored on the execution itself. Then when calculating if we still have budget, use these estimates as the base.
> This would allow us later on to write the actual costs (after execution) into that table instead of using the estimate.
> Rewriting the estimated vs. the actual costs would give us another problem though: reproducability and debugging issues. We wouldn't know what the system did unless we have a good logging.
> Maybe instead we should record for each execution in a separate table the estimates, then the actual costs, and what else is needed..
> In any case, we should discuss this more before merging it IMO.

Definitely!

**Comment by cbuchli** [comment]
I'll have to re-open this one because it also fixes a few bugs in the estimation. I'll remove the "blocking" and visible parts.

**Comment by mweibel** [comment]
fair enough. gonna merge. 
What I didn't think of yet is the `isCharity` flag, which also goes into that calculation I assume. 
But yeah :) Things to keep in mind for next iterations.

File: src/Classes/Controller/ApiJobExecuteController.php [hunk 8]
```diff
@@ -173,35 +174,47 @@ public function execAction(): ResponseInterface
...
+                $result['estimates'] = JobFactory::getDispatchConfigOfJob($this->job, $this->execution)->getExecutionEstimates();
+            } else {
+                $this->persistJob();
+            }
+
+            return $this->render($result);
         } catch (Exception $e) {
             return $this->render(['success' => false, 'message' => $e->getMessage()], StatusCode::HTTP_INTERNAL_SERVER_ERROR);
+        } catch (GuzzleException $e) {
+            return $this->render(['success' => false, 'message' => $e->getMessage()], StatusCode::HTTP_INTERNAL_SERVER_ERROR);
```
Comment (mweibel) [inline src/Classes/Controller/ApiJobExecuteController.php]: actually, why is that here? the above catch should catch this already, no?

Comment (cbuchli) [reply src/Classes/Controller/ApiJobExecuteController.php]: :D good catch, PHPStorm has added this automatically and I didn't look properly. Probably, we should switch to Throwable anyways.
